### DISCUSSION TITLE
how to convert integer (number) to byte in Lua

### DISCUSSION THREAD

**Discussion by bijlevel** [RomanKubiak/ctrlr#584] [Q&A] [answered]
Just what the title says: how do I convert a integer to a byte? I came across the string.byte() function and maybe that's the way to go: first convert the integer to a string  and then use string.byte() function? This way it seems to me a  bit of a detour...

**Comment by dnaldoog** [comment]
Hi @bijlevel 

An integer **is** a byte or a series of bytes if the integer is over 255. 
[_i.e 256 = 01 00_ is a 2 byte number. `0x100`]

A byte has a limit of 8 bits, which as an integer has a maximum value of  255 or "FF" as a binary string or `1111 1111` as binary. FF can also represent -1 in 8 bit signed integer.

ASCII text has a range in the 8 bits from 32 ~ 126 for text so depending how you interpret that byte, it could represent a string or a number 
- ie 32 could be the number 32 (0x20) or a space character. It's up to you and how you interpret the byte.

### Here decimal 65 = 0x41

- console(String(string.byte("A"))) -- will print 65
- console(string.char(65)) -- will print "A"
- console(MemoryBlock("41"):toString()) -- will print A
- console(MemoryBlock("41"):toHexString(1)) -- will print 41
- console(String(MemoryBlock("41"):getByte(0))) -- will print 65
- console(string.format("%.2x",65))  -- will print 41
- console(bit.tohex(65,-2)) -- will print 41

With JUCE MemoryBlock in Ctrlr you cannot define a MemoryBlock() with an integer - as in `m=MemoryBlock(65)`; you have to have a binary string or lua table as initializers.

But in all of these examples all the computer 'sees' is a set of 8 ~ 0s or 1s `0100 0001`

 `m=MemoryBlock(65)` -- will **not** work
 `m=MemoryBlock("41")` -- will work
 `m=MemoryBlock({65})` -- will work

----
[binary strings etc_1_0.zip](https://github.com/RomanKubiak/ctrlr/files/13692565/binary.strings.etc_1_0.zip)

**Reply by bijlevel** [reply]
Thanks for your extensive answer, more than I deserve ;-).  I think in my case **bit.tohex(65,-2**) will do, be it that I have to replace 65 with a variable. But then I have to split up the byte into 2 nibbles. I thought I had the right code for this:

`t1 = data:getByte(i)
t2 = BigInteger(t1)
            t3 = data:getByte(i+1)
            t4 = BigInteger(t3)
            nib1 = t2:getBitRangeAsInt(0,4)
            nib2 = t4:getBitRangeAsInt(0,4)`
In this example t1 (and t2) are for sure bytes, because they come from received midi data, but when I use this on the result of the bit.tohex function it gives me an error (or even worse, CTRLR crashes...) So how to do this: split up a byte into two nibbles?

While writing this post your next post came in. Thanks again, but I know how to insert a value in a sysex string. That would be, in my case, the next step.

Edit: I realise now that the function bit.tohex() does just what it says, covert to hex, but that's of course not a byte...

I tried the MemoryBlock ({65}) example, but the result is also not a byte?

**Comment by dnaldoog** [comment]
Perhaps you are trying to add a value from a `uiSlider` etc to a sysex string. That value is an integer of course, but you can easily insert it into a sysex message (set of binary strings) like this:

<pre>
f=function(mod,value,source)
local sysex=string.format("F0 00 00 00 %.2x F7",value)
-- send sysex
panel:sendMidiMessageNow(CtrlrMidiMessage(sysex))
</pre>

Sometimes I also do this:

<pre>
f=function(mod,value,source)
local sysex=MemoryBlock("F0 00 00 00 00 F7")
sysex:setByte(4,value)
-- send sysex
panel:sendMidiMessageNow(CtrlrMidiMessage(sysex:toHexString(1)))
</pre>

**Comment by dnaldoog** [comment, marked as answer]
Hi @bijlevel 

To split a byte into two four bit nibbles you can use this function. Lua can return two values from a function:
<pre>
function create4n(n) -- convert integer into two 4 bit nibbles
    local bv = BigInteger(n)
    return bv:getBitRangeAsInt(4, 4), bv:getBitRangeAsInt(0, 4)
end
</pre>

create your two variables:

`local msb,lsb=create4n(value)`

should do the trick ... then:
<pre>
local sysex=string.format("F0 00 00 %.2x %.2x F7",msb,lsb)
-- send sysex
panel:sendMidiMessageNow(CtrlrMidiMessage(sysex))
</pre>

----

__P.S.__ You can do the reverse with:

<pre>
function combine4n(a, b) -- convert two 4 bit nibbles to a single integer
    local m = bit.lshift(a, 4)
    return m + b
end 

--[[ or --]]

function combine4n(a, b) -- convert two 4 bit nibbles to a single integer
    return bit.bor(bit.lshift(a, 4), b)
end


local valueForMod=combine4n(m:getByte(4),m:getByte(5))
</pre>

**Reply by bijlevel** [reply]
Thanks! Your code works! And you're right, for the computer everything is 0s and 1s, so then only "conversion" thing needed in this case is the use of BigInteger.
